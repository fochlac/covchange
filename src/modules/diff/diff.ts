export function diffReports(original: Core.Report, changed: Core.Report): Core.DiffReport {
	const deletedFiles = Object.keys(original.files).filter(name => !changed.files[name])
	const changedFiles = Object.keys(original.files).filter(name => {
		const altered = changed.files[name] && changed.files[name].metrics
		const orig = original.files[name].metrics
		if (!altered) {
			return false
		}

		return altered && (altered.statements !== orig.statements || altered.coveredStatements !== orig.coveredStatements)
	})
	const newFiles = Object.keys(changed.files).filter(name => !original.files[name])

	return {
		total: diffMetrics(original.metrics, changed.metrics),
		deleted: deletedFiles,
		changed: changedFiles.reduce((files, name) => {
			files[name] = diffMetrics(original.files[name].metrics, changed.files[name].metrics)
			return files
		}, {}),
		new: newFiles.reduce((files, name) => {
			files[name] = changed.files[name].metrics
			return files
		}, {}),
	}
}

function diffMetrics(original: Core.Metrics, changed: Core.Metrics): Core.Diff {
	return {
		original,
		changed,
		diff: {
			statements: Math.floor((changed.statements - original.statements) * 10) / 10,
			conditionals: Math.floor((changed.conditionals - original.conditionals) * 10) / 10,
			coveredStatements: Math.floor((changed.coveredStatements - original.coveredStatements) * 10) / 10,
			coveredConditionals: Math.floor((changed.coveredConditionals - original.coveredConditionals) * 10) / 10,
			statementCov: Math.floor((changed.statementCov - original.statementCov) * 10) / 10,
			conditionalCov: Math.floor((changed.conditionalCov - original.conditionalCov) * 10) / 10,
		},
	}
}
