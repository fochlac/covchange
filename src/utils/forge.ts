export const forge = {
	repositoryId,
	branchId,
	pullRequestId,
}

function repositoryId({ name, slug }: Core.Repository): Core.Id {
	return `${slug.replace('/', '_')}_${name}`
}

function branchId({ name, repository }: Core.Branch | Core.BranchRest | { name: string; repository: Core.Repository }): Core.Id {
	return `${repositoryId(repository)}_${name}`
}

function pullRequestId({ pullRequestId, repository }: { pullRequestId: string; repository: Core.Repository }): Core.Id {
	return `${repositoryId(repository)}_${pullRequestId}`
}
