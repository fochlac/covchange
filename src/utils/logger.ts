import { createWriteStream, existsSync, mkdirSync } from 'fs'

/**
 *   Log Levels:
 *   0: no error, very important info
 *   1: error: critical error in app
 *   2: error: internal error in app
 *   3: error: minor error
 *
 *   4: info: important info (bad request/login)
 *   5: info: minor info
 *
 *   6: trace: tracing informations
 *   7: unimportant tracing information
 *
 *  10: full trace: (requests objects, etc)
 *
 **/
let currentDate = new Date().toISOString().split('T')[0]
if (!existsSync(global.logFolder)) {
	mkdirSync(global.logFolder)
}

let logStream = createWriteStream(global.logFolder + '/covwatcher_' + currentDate + '.log', {
	flags: 'a',
	encoding: 'utf8',
	autoClose: true,
})

logStream.on('error', err => {
	console.log(err)
})

export default (level: number, ...message: any[]): void => {
	const now = new Date().toISOString()
	const messages = message.map(parseMessage).join(' ')
	const logMessage = `${now} - ${level} - ${messages}`

	if (level <= global.logLevel) {
		console.log(logMessage)
	}

	if (currentDate !== new Date().toISOString().split('T')[0]) {
		currentDate = new Date().toISOString().split('T')[0]
		logStream.end()
		logStream = createWriteStream(global.logFolder + '/covwatcher_' + currentDate + '.log', {
			flags: 'a',
			encoding: 'utf8',
			autoClose: true,
		})
	}

	logStream.write(logMessage + '\n')
}

function parseMessage(item) {
	let output
	if (['string', 'number', 'boolean'].includes(typeof item)) {
		output = item
	} else {
		try {
			if (Object.prototype.toString.call(item) === '[object Error]') {
				output = '\n' + item.stack.toString()
			} else {
				output = '\n' + JSON.stringify(item)
			}
		} catch (err) {
			console.log(err + '\n' + item)
			output = ''
		}
	}
	return output
}
