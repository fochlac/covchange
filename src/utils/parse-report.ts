import { parse } from 'fast-xml-parser'

const config = {
	ignoreAttributes: false,
	attrNodeName: 'attr',
	attributeNamePrefix: '',
	parseAttributeValue: true,
}

/**
 * Function to parse a raw clover coverage report in xml format
 *
 * @param {string} rawXML coverage report as utf8 string
 * @returns {Core.Report} parsed coverage report
 */
export function createReportFromXML(rawXML: string): Core.Report {
	const rawProject = parse(rawXML, config).coverage.project

	const folders = rawProject.metrics.package.reduce(reduceFolder, {})
	const files = Object.values(folders).reduce(reduceFiles, {})

	return {
		timestamp: rawProject.attr.timestamp,
		metrics: parseMetrics(rawProject.metrics.attr),
		folders,
		files,
	}
}

function reduceFolder(folders, folder) {
	folders[folder.attr.name] = {
		name: folder.attr.name,
		metrics: parseMetrics(folder.metrics.attr),
		files: parseFile(folder.file),
	}
	return folders
}

function reduceFiles(files, folder) {
	folder.files.forEach(file => {
		files[`${folder.name.split('.').join('/')}/${file.name}`] = file
	})
	return files
}

function parseFile(file) {
	const rawFiles = file && file.length ? file : file ? [file] : []

	return rawFiles.map(file => ({
		name: file.attr.name,
		metrics: parseMetrics(file.metrics.attr),
	}))
}

function parseMetrics({ statements, coveredstatements, conditionals, coveredconditionals }) {
	return {
		statements,
		conditionals,
		coveredStatements: coveredstatements,
		coveredConditionals: coveredconditionals,
		statementCov: (statements && Math.round((coveredstatements / statements) * 10000)) / 100 || 0,
		conditionalCov: (conditionals && Math.round((coveredconditionals / conditionals) * 10000)) / 100 || 0,
	}
}
