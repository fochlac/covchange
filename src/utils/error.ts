import logger from './logger'

const log = (level, ...message) => logger(level, 'error -', ...message)

const error = (location: string): Core.ErrorConstructor => ({
	internalError: (level: number, ...message: Array<any>) => (detail: Error): void => {
		log(level, location, ...message, detail)
	},
	routerError: (level: number, res: Express.Response, ...message: Array<any>) => (detail: Error): void => {
		log(level, location, ...message, detail)

		res.status(500).send({ success: false })
	},
	badRequest: (level: number, res: Express.Response, field: string, ...message: Array<any>) => (detail: Error): void => {
		log(level, location, ...message, detail)

		res.status(403).send({ success: false, field })
	}
})

export default error
