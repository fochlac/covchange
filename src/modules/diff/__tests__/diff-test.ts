import { createReportFromXML } from '../../../utils/parse-report'
import { diffReports } from '../diff'
import { readFileSync } from 'fs'

describe('diff-module', () => {
	const report = createReportFromXML(readFileSync('./tests/sample_report.xml', { encoding: 'utf8' }))
	const report2 = createReportFromXML(readFileSync('./tests/sample_report_2.xml', { encoding: 'utf8' }))

	it('should find the diff between the reports', () => {
		expect(diffReports(report, report2)).toMatchSnapshot()
	})
})
