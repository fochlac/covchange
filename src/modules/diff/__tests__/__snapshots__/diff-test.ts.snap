// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff-module should find the diff between the reports 1`] = `
Object {
  "changed": Object {
    "orders/app/components/order-tours-tab/order-tours-table-renderer.js": Object {
      "changed": Object {
        "conditionalCov": 31.4,
        "conditionals": 35,
        "statementCov": 44.4,
        "statements": 45,
      },
      "diff": Object {
        "conditionalCov": 31.4,
        "conditionals": 0,
        "statementCov": 44.4,
        "statements": 10,
      },
      "original": Object {
        "conditionalCov": 0,
        "conditionals": 35,
        "statementCov": 0,
        "statements": 35,
      },
    },
    "orders/app/components/orders-filter-bar.js": Object {
      "changed": Object {
        "conditionalCov": 50,
        "conditionals": 22,
        "statementCov": 96.4,
        "statements": 55,
      },
      "diff": Object {
        "conditionalCov": 18.2,
        "conditionals": 0,
        "statementCov": 12.8,
        "statements": -12,
      },
      "original": Object {
        "conditionalCov": 31.8,
        "conditionals": 22,
        "statementCov": 83.6,
        "statements": 67,
      },
    },
    "orders/app/components/orders-subviews.js": Object {
      "changed": Object {
        "conditionalCov": 94.7,
        "conditionals": 19,
        "statementCov": 68,
        "statements": 25,
      },
      "diff": Object {
        "conditionalCov": -5.3,
        "conditionals": 2,
        "statementCov": -13,
        "statements": 4,
      },
      "original": Object {
        "conditionalCov": 100,
        "conditionals": 17,
        "statementCov": 81,
        "statements": 21,
      },
    },
    "orders/app/index.js": Object {
      "changed": Object {
        "conditionalCov": 0,
        "conditionals": 4,
        "statementCov": 7.7,
        "statements": 13,
      },
      "diff": Object {
        "conditionalCov": 0,
        "conditionals": 0,
        "statementCov": 7.7,
        "statements": 1,
      },
      "original": Object {
        "conditionalCov": 0,
        "conditionals": 4,
        "statementCov": 0,
        "statements": 12,
      },
    },
  },
  "deleted": Array [
    "orders/module.js",
  ],
  "new": Object {
    "orders/app/components/order-tours-tab/order-tours-table-core.js": Object {
      "conditionalCov": 71.4,
      "conditionals": 7,
      "statementCov": 54.5,
      "statements": 11,
    },
  },
  "total": Object {
    "changed": Object {
      "conditionalCov": 80.2,
      "conditionals": 2489,
      "statementCov": 87.2,
      "statements": 4082,
    },
    "diff": Object {
      "conditionalCov": 7.2,
      "conditionals": 10,
      "statementCov": 2.4,
      "statements": 40,
    },
    "original": Object {
      "conditionalCov": 72.9,
      "conditionals": 2479,
      "statementCov": 84.8,
      "statements": 4042,
    },
  },
}
`;
